### ISSUE TITLE
How to add dynamically defined module and its optimizer to the lightning module?

### ISSUE THREAD

**Issue by showgood163** [PyTorchLightning/pytorch-lightning#4590]
#### What is your question?

I have a module whose configuration is defined by one forward pass, e.g. a fc layer whose input size depends on some random features inside a network. I also have to update the parameters of this module. What's the recommend way to do this?

#### Code

[PatchSampleF is a kind of this module](https://github.com/taesungp/contrastive-unpaired-translation/blob/afdc8fb02742b4eef7724ca452fcee38016ee008/models/networks.py#L531), 

#### What have you tried?

I try to achieve this by defining the module and its optimizer inside the first validation step ( pl.Trainer(num_sanity_val_steps=1) ) before training. However, in the following training step, all parameters of the newly-defined module do not requires grad anymore. A workaround is to manually require_grad for these parameters. I don't know what triggered the change but I think if there is a way to register the newly-defined module and corresponding optimizer to the training process, this problem can be solved gracefully.

#### What's your environment?

 - OS: Arch linux
 - Packaging pip
 - Version 1.0.5

**Comment by tchaton** [comment]
Hey @showgood163,

I would say the simplest approach is the following:
* you use Trainer(automatic_optimization=False) and in your training_step, you make your forward (which will define your model), your create / override your optimiser there, and perform manual_backward, manual_optimizer_step with your dynamically defined optimiser.
I think it should work fine :)

Here is the doc: https://pytorch-lightning.readthedocs.io/en/stable/_modules/pytorch_lightning/core/lightning.html#LightningModule.manual_backward

Best regards,
Thomas Chaton.

**Comment by showgood163** [comment]
Thanks! Will have a try and update if I have a result.

**Comment by showgood163** [comment]
I'm encountering problems about the convergence of the model, even I'm running the code in the single gpu mode.

- I tried `automatic_optimization=False` and change the code according to [this](https://pytorch-lightning.readthedocs.io/en/1.0.5/optimizers.html#manual-optimization). After that I found that all the logs using pl.module.log() is gone and I fallback to manual logging methods. So what should I do to make the automatic logging work?

- After changing the code, I found that the traning doesn't converge. I also tested on a non-lightning version of my code and it does converge. In addition, I tested manual_backward(), step(), zero_grad() to make sure it works. So what could possibly go wrong here?

- manual_optimizer_step is not in the released 1.0.5 version, right?
